update_copyright.d

```diff
@@ -0,0 +1,371 @@
+
+    if (args.length == 1 || opts.helpWanted)
+    {
+        defaultGetoptPrinter("usage: update_copyright [--help|-h] [--update-year|-y] <dir>...",
+                             opts.options);
+        return 0;
+    }
+
+    Copyright(updateYear, verbose).run(args[1 .. $]);
+    return 0;
+}
+
+//::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
+
+struct Copyright
+{
+    import std.algorithm, std.array, std.file, std.regex, std.stdio, std.string;
+
+    // The author to use in copyright notices.
+    enum author = "The D Language Foundation, All Rights Reserved";
+
+    // The standard (C) form.
+    enum copyright = "(C)";
+
+private:
+    // True if running in verbose mode.
+    bool verbose = false;
+
+    // True if also updating copyright year.
+    bool updateYear = false;
+
+    // An associative array of known copyright holders.
+    // Value set is true if the copyright holder is internal.
+    bool[string] holders;
+
+    // Files and directories to ignore during search.
+    static string[] skipDirs = [
+        "docs",
+        "ini",
+        "test",
+        "samples",
+        "vcbuild",
+        ".git",
+    ];
+
+    static string[] skipFiles = [
+        "Jenkinsfile",
+        "LICENSE.txt",
+        "VERSION",
+        ".a",
+        ".ddoc",
+        ".deps",
+        ".lst",
+        ".map",
+        ".md",
+        ".o",
+        ".obj",
+        ".sdl",
+        ".sh",
+        ".yml",
+    ];
+
+    // Characters in a range of years.
+    // Include '.' for typos, and '?' for unknown years.
+    enum rangesStr = `[0-9?](?:[-0-9.,\s]|\s+and\s+)*[0-9]`;
+
+    // Non-whitespace characters in a copyright holder's name.
+    enum nameStr = `[\w.,-]`;
+
+    // Matches a full copyright notice:
+    // - 'Copyright (C)', etc.
+    // - The years. Includes the whitespace in the year, so that we can
+    //   remove any excess.
+    // - 'by ', if used
+    // - The copyright holder.
+    Regex!char copyrightRe;
+
+    // A regexp for notices that might have slipped by.
+    Regex!char otherCopyrightRe;
+
+    // A regexp that matches one year.
+    Regex!char yearRe;
+
+    // Matches part of a year or copyright holder.
+    Regex!char continuationRe;
+
+    Regex!char commentRe;
+
+    // Convenience for passing around file/line number information.
+    struct FileLocation
+    {
+        string filename;
+        size_t linnum;
+
+        string toString()
+        {
+            import std.format;
+            return "%s(%d)".format(this.filename, this.linnum);
+        }
+    }
+
+    void processFile(string file)
+    {
+        // Looks like something we tried to create before.
+        if (file.endsWith(".tmp"))
+        {
+            remove(file);
+            return;
+        }
+
+        auto h = File(file, "rb");
+        auto lines = appender!string;
+        int errors = 0;
+        bool changed = false;
+
+        lines.reserve(h.size);
+
+        char[] orig_line = null;
```

Comment:
*NIT
casing seems inconsistent. Maybe this was ported from python, and you have some of the original casing left over from the port?

Reply:
No, I usually follow a convention to distinguish between local vars, field, and methods.  It makes it easier to infer where something comes from, but can go with camelCase if that really makes you happy.

Reply:
Sounds like a reasonable convention but doesn't match any convention I've seen in `dmd/druntime/phobos/tools`.  But casing isn't a deal breaker for me.  Do what you think is best.